spring-cloud-azure-keyvault-config/src/main/java/com/microsoft/azure/spring/cloud/keyvault/config/KeyVaultPropertySourceLocator.java

```diff
@@ -8,22 +8,92 @@
+import org.springframework.core.env.CompositePropertySource;
+import org.springframework.core.env.ConfigurableEnvironment;
 import org.springframework.core.env.Environment;
 import org.springframework.core.env.PropertySource;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
 
 /**
  * Builds a {@link KeyVaultPropertySource} instance based on application name and active profiles.
  */
 @AllArgsConstructor
+@Slf4j
 public class KeyVaultPropertySourceLocator implements PropertySourceLocator {
+    private static final String DEFAULT_APP_NAME_PROPERTY = "spring.application.name";
+    private static final String PROPERTY_SOURCE_NAME = "azure-key-vault-config";
+    // TODO: support Azure China, Azure Germany, Azure Government
+    private static final String GLOBAL_KEY_VAULT_ENDPOINT_TEMPLATE = "https://%s.vault.azure.net";
+
+    private static final String FAILED_TO_LOAD = "Failed to load configuration from Azure Key Vault: ";
 
     private final KeyVaultClient keyVaultClient;
 
     private final KeyVaultConfigProperties properties;
 
     @Override
     public PropertySource<?> locate(Environment environment) {
-        return null;
+        if (!(environment instanceof ConfigurableEnvironment)) {
+            return null;
+        }
+
+        if (!properties.isEnabled()) {
+            return null;
+        }
+
+        final List<String> keyVaults = buildKeyVaultUrl((ConfigurableEnvironment) environment);
+
+        CompositePropertySource composite = new CompositePropertySource(PROPERTY_SOURCE_NAME);
+        for (String keyVault : keyVaults) {
+            try {
+                composite.addPropertySource(new KeyVaultPropertySource(keyVault, keyVaultClient, keyVault));
+            } catch (Exception ex) {
+                if (this.properties.isFailFast()) {
+                    log.error(FAILED_TO_LOAD + keyVault);
+                    ReflectionUtils.rethrowRuntimeException(ex);
+                } else {
+                    log.warn(FAILED_TO_LOAD + keyVault, ex);
+                }
+            }
+        }
+
+        return composite;
+    }
+
+    // Key Vault name has the following formats:
+    // - "{app name}-{profile}" when both of them are present
+    // - "{app name}" when profile is empty
+    private List<String> buildKeyVaultUrl(ConfigurableEnvironment environment) {
+
+        String appName = properties.getName();
+        if (StringUtils.isEmpty(appName)) {
+            appName = environment.getProperty(DEFAULT_APP_NAME_PROPERTY);
+        }
+
+        String[] profiles = StringUtils.commaDelimitedListToStringArray(properties.getActiveProfiles());
+        if (profiles == null || profiles.length == 0) {
+            profiles = environment.getActiveProfiles();
+        }
+
+        if (profiles == null || profiles.length == 0) {

```

Comment:
if condifition is the same as line 81 ?

Reply:
Yes, checking against different values. See line 82.

Reply:
OK. I see, personally I would like to move the second checking after line 82, more native I think.